nitorch/tools/affine_reg.py

```diff
@@ -0,0 +1,961 @@
+    fix : int, default=0
+        Index of image to used as fixed image, not used if mean_space=True.
+    fov : (2,) tuple, default=None
+        A tuple with affine matrix (tensor_like) and dimensions (tuple) of mean space.
+    verbose : bool, default=False
+        Show registration results.
+    device : torch.device or str, default='cpu'
+        PyTorch device type.
+
+    Returns
+    ----------
+    q : (N, Nq) tensor_like (N=number of input images, Nq=number affine bases)
+        Lie algebra of affine registration fit. Get matrix representation by:
+            A = dexpm(q[n, ...], B)[0]
+    M_fix : (4, 4), tensor_like
+        Affine matrix of fixed image.
+    dim_fix : (3,), tuple, list, tensor_like
+        Image dimensions of fixed image.
+    B : (4, 4, Nq), tensor_like
+        Lie basis.
+
+    """
+    # Parse algorithm options
+    opt = {'optimiser': optimiser,
+           'cost_fun': cost_fun,
+           'device': device,
+           'samp': samp,
+           'fix': fix,
+           'mean_space': mean_space,
+           'verbose': verbose,
+           'fov': fov,
+           'basis' : basis}
+
+    # Some very basic sanity checks
+    N = len(imgs) # Number of input scans
+    mov = list(range(N))  # Indices of images
+    if N < 2:
+        raise ValueError('At least two images needed!')
+    if opt['cost_fun'] in costs_hist and opt['mean_space']:
+        raise ValueError('Option mean_space=True not defined for {} cost!'.format(opt['cost_fun']))
+
+    # Torch stuff
+    opt = _init_torch(opt)
+
+    # Rigid registration basis
+    B = get_affine_basis(basis=opt['basis'], device=opt['device'])
+    Nq = B.shape[-1]
+
+    # Load data
+    dat, mats, dims = _data_loader(imgs, opt)
+
+    # Define fixed image space (M_fix, dim_fix, vx_fix)
+    if opt['mean_space']:
+        # Use a mean-space
+        if opt['fov']:
+            # Mean-space given
+            M_fix = opt['fov'][0]
+            dim_fix = opt['fov'][1]
+        else:
+            # Compute mean-space
+            M_fix, dim_fix = _get_mean_space(mats, dims)
+        arg_dat_grid = dim_fix
+    else:
+        # Use one of the input images
+        M_fix = mats[opt['fix']]
+        dim_fix = dat[opt['fix']].shape[:3]
+        mov.remove(opt['fix'])
+        arg_dat_grid = dat[opt['fix']]
+    # Get voxel size of fixed image
+    vx_fix = voxel_size(M_fix)
+
+    # Initial guess for registration parameter
+    q = torch.zeros((N, Nq), dtype=torch.float32)
+    if opt['optimiser'] == 'powell':
+        # SciPy's powell optimiser requires Numpy arrays
+        q = q.numpy()
+
+    # Do registration
+    fig_ax = None  # Just used for visualisation
+    for s in opt['samp']:  # Loop over sub-sampling level
+        # Get possibly sub-sampled fixed image, and its interpolation grid
+        dat_fix, grid = _get_dat_grid(arg_dat_grid, vx_fix, s, opt)
+        # Do optimisation
+        if opt['cost_fun'] == 'njtv':  # Groupwise optimisation
```

Comment:
I don't think I understand the question? The difference between mean_space True or False is that if it is True then all images will be considered moving and matched in a mean space, otherwise one will be picked as fixed and the other images mixed to this fixed image:

``` py

'''
fix : int, default=0
    Index of image to used as fixed image, not used if mean_space=True.
'''
```

I was thinking that using a fixed image could be nice with NJTV to match to an MNI space maybe?

Reply:
I think my question is: is the optimisation of njtv groupwise even when mean_space is False?

Reply:
Yes, then it uses NJTV with one image fixed:

``` py
'''
fix : int, default=0
'''
```